drivers/mbox/mbox_esp32.c

```diff
@@ -0,0 +1,267 @@
+#include "soc/gpio_periph.h"
+
+#include <stdint.h>
+#include <string.h>
+#include <zephyr/device.h>
+#include <zephyr/drivers/mbox.h>
+#include <zephyr/drivers/interrupt_controller/intc_esp32.h>
+#include <soc.h>
+#include <zephyr/sys/atomic.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(mbox_esp32, CONFIG_MBOX_LOG_LEVEL);
+
+#define ESP32_MBOX_LOCK_FREE_VAL 0xB33FFFFF
+#define ESP32_MBOX_NOOP_VAL 0xFF
+
+__packed struct esp32_mbox_control {
+	uint16_t dest_cpu_msg_id[2];
+	atomic_val_t lock;
+};
+
+struct esp32_mbox_memory {
+	uint8_t *pro_cpu_shm;
+	uint8_t *app_cpu_shm;
+};
+
+struct esp32_mbox_config {
+	uint32_t irq_source_pro_cpu;
+	uint32_t irq_source_app_cpu;
+};
+
+struct esp32_mbox_data {
+	mbox_callback_t cb;
+	void *user_data;
+	uint32_t this_core_id;
+	uint32_t other_core_id;
+	uint32_t shm_size;
+	struct esp32_mbox_memory shm;
+	struct esp32_mbox_control *control;
+};
+
+IRAM_ATTR static void esp32_mbox_isr(const struct device *dev)
+{
+	struct esp32_mbox_data *dev_data = (struct esp32_mbox_data *)dev->data;
+	uint32_t core_id = dev_data->this_core_id;
+
+	/* clear interrupt flag */
+	if (core_id == 0) {
+#if defined(CONFIG_SOC_SERIES_ESP32)
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_0_REG, 0);
+#elif defined(CONFIG_SOC_SERIES_ESP32S3)
+		WRITE_PERI_REG(SYSTEM_CPU_INTR_FROM_CPU_0_REG, 0);
+#endif
+	} else {
+#if defined(CONFIG_SOC_SERIES_ESP32)
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_1_REG, 0);
+#elif defined(CONFIG_SOC_SERIES_ESP32S3)
+		WRITE_PERI_REG(SYSTEM_CPU_INTR_FROM_CPU_1_REG, 0);
+#endif
+	}
+
+	/* first of all take the own of the shared memory */
+	while (!atomic_cas(&dev_data->control->lock,
+		ESP32_MBOX_LOCK_FREE_VAL, dev_data->this_core_id))
+		;
+
+	if (dev_data->cb) {
+
+		volatile void *shm = dev_data->shm.pro_cpu_shm;
+
+		if (core_id != 0) {
+			shm = dev_data->shm.app_cpu_shm;
+		}
+	}
+
+	/* unlock the shared memory */
+	atomic_set(&dev_data->control->lock, ESP32_MBOX_LOCK_FREE_VAL);
+}
+
+static int esp32_mbox_send(const struct device *dev, mbox_channel_id_t channel,
+			    const struct mbox_msg *msg)
+{
+	ARG_UNUSED(msg);
+
+	struct esp32_mbox_data *dev_data = (struct esp32_mbox_data *)dev->data;
+
+	if (channel > 0xFFFF) {
+		LOG_ERR("Invalid channel");
+		return -EINVAL;
+	}
+
+	uint32_t key = irq_lock();
+
+	/* try to lock the shared memory */
+	while (!atomic_cas(&dev_data->control->lock,
+		ESP32_MBOX_LOCK_FREE_VAL,
+		dev_data->this_core_id)) {
+		k_msleep(1);
+	}
+
+	/* Only the lower 16bits of id are used */
+	dev_data->control->dest_cpu_msg_id[dev_data->other_core_id] = (uint16_t)(channel & 0xFFFF);
+
+	/* data copied, set the id and, generate interrupt in the remote core */
+	if (dev_data->this_core_id == 0) {
+		atomic_set(&dev_data->control->lock, ESP32_MBOX_LOCK_FREE_VAL);
+		LOG_DBG("Generating interrupt on remote CPU 1 from CPU 0");
+#if defined(CONFIG_SOC_SERIES_ESP32)
+		DPORT_WRITE_PERI_REG(DPORT_CPU_INTR_FROM_CPU_1_REG, DPORT_CPU_INTR_FROM_CPU_1);
```

Comment:
Sounds you mixed the interrupt generation? In here, it should cause interrupt from CPU 0 instead: `DPORT_CPU_INTR_FROM_CPU_0`. Same for the `else` condition..

Reply:
It is the same code from the working IPM driver, so no differences here.

Please check:

https://github.com/zephyrproject-rtos/zephyr/blob/6211de87c0660a5dfd21b5decc93a67236ec0984/drivers/ipm/ipm_esp32.c#L138

Reply:
It is correct because when you send a mailbox signal, the current CPU might want to generate interrupt signal in the other core.

Using the CPU_0 software interrupt register generate the opposite behavior, so the current CPU would generate a SW interrupt for itself which is not intended.